devito/operations/interpolators.py

```diff
@@ -388,3 +389,520 @@ def callback():
+        index_matrix, points = self.sfunction._index_matrix(offset)
+
+        idx_subs = []
+        for i, idx in enumerate(index_matrix):
+            # Introduce ConditionalDimension so that we don't go OOB
+            mapper = {}
+            for j, d in zip(idx, self.grid.dimensions):
+                p = points[j]
+                lb = sympy.And(p >= d.symbolic_min - self.sfunction._radius,
+                               evaluate=False)
+                ub = sympy.And(p <= d.symbolic_max + self.sfunction._radius,
+                               evaluate=False)
+                condition = sympy.And(lb, ub, evaluate=False)
+                mapper[d] = ConditionalDimension(p.name, self.sfunction._sparse_dim,
+                                                 condition=condition, indirect=True)
+
+            # Track Indexed substitutions
+            idx_subs.append(mapper)
+
+        # Temporaries for the position
+        temps = [Eq(v, k, implicit_dims=self.sfunction.dimensions)
+                 for k, v in self.sfunction._position_map.items()]
+
+        temps.extend([Eq(v, k.subs(self.sfunction._position_map),
+                      implicit_dims=self.sfunction.dimensions)
+                     for k, v in self.sfunction._relative_position_map.items()])
+
+        # Temporaries for the indirection dimensions
+        temps.extend([Eq(v, k.subs(self.sfunction._position_map),
+                         implicit_dims=self.sfunction.dimensions)
+                      for k, v in points.items()])
+
+        return idx_subs, temps
+
+    def interpolate(self, expr, offset=0, increment=False, self_subs={}):
+        """
+        Generate equations interpolating an arbitrary expression into ``self``.
+        Parameters
+        ----------
+        expr : expr-like
+            Input expression to interpolate.
+        offset : int, optional
+            Additional offset from the boundary.
+        increment: bool, optional
+            If True, generate increments (Inc) rather than assignments (Eq).
+        """
+        def callback():
+
+            # Derivatives must be evaluated before the introduction of indirect accesses
+            try:
+                _expr = expr.evaluate
+            except AttributeError:
+                # E.g., a generic SymPy expression or a number
+                _expr = expr
+
+            variables = list(retrieve_function_carriers(_expr))
+
+            # Need to get origin of the field in case it is staggered
+            # TODO: handle each variable staggereing spearately
+            field_offset = variables[0].origin
+
+            # List of indirection indices for all adjacent grid points
+            idx_subs, temps = self._interpolation_indices(variables, offset,
+                                                          field_offset=field_offset)
+
+            rhs = Symbol(name='sum', dtype=self.sfunction.dtype)
+            summands = [Eq(rhs, 0., implicit_dims=self.sfunction.dimensions)]
```

Comment:
everything till here is the same as in `LinearInterpolation` right? and after the `bicubic_equation` too. So may be nice to gather those under a `PolynomialInterpolation` with `order=1` or `order=3` that specialize on the equations only since it's the only thing that changes

Reply:
I added a PolynomialInterpolator class that works as a parent class. It's not exactly what you just said, but I think it solves the code reuse problem